add ability to specify custom DNS settings for a VM in the cluster

nawazkh opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind feature

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

**What this PR does / why we need it**:
- A user would be able to specify a custom DNS setting for a VM in the cluster.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2182 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
AzureMachineSpec has been updated with DNSServers field. Users get to specify custom DNS Settings for a VM in the cluster.
```

invidian commented:
/ok-to-test

mboersma commented:
/milestone v1.5

nawazkh commented:
/retitle [WIP] NIC specific DNS settings

nawazkh commented on azure/services/networkinterfaces/spec.go:
```diff
@@ -125,6 +126,11 @@ func (s *NICSpec) Parameters(existing interface{}) (parameters interface{}, err
 		s.AcceleratedNetworking = &accelNet
 	}
 
+	dnsSettings := network.InterfaceDNSSettings{}
+	if len(s.DNSServers) != 0 {
+		dnsSettings.DNSServers = &s.DNSServers
+	}
+
```
@jsturtevant Is it okay to be passing `<nil>` `dnsSettings` in case `s.DNSServers` list is empty?

jackfrancis replied:
Are you concerned about a nil slice being evaluated as `!=  0`? In fact that's not the case in golang. See:

https://go.dev/play/p/E0ilKGlGiMg

nawazkh replied:
Sorry that I confused, I meant to ask https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2447/files#r932527885

jsturtevant replied:
I think you got your answer :-)  this looks ok to me....

nawazkh commented on azure/services/networkinterfaces/spec.go:
```diff
@@ -150,6 +156,7 @@ func (s *NICSpec) Parameters(existing interface{}) (parameters interface{}, err
 		InterfacePropertiesFormat: &network.InterfacePropertiesFormat{
 			EnableAcceleratedNetworking: s.AcceleratedNetworking,
 			IPConfigurations:            &ipConfigurations,
+			DNSSettings:                 &dnsSettings,
```
Sorry to confuse, please ignore, I think I answered my own question as I was thinking over the returned `network.Interface` object.

I had meant to ask "if we are going to break anything later on if `&dnsSettings` is a `nil` struct in the below return type? 
Should we be appending an empty slice instead to `&dnsSettings.DNSServers` ?"

But I realized that `&dnsSettings` can be nil if `len(s.DNSServers) == 0`. 

I.e. below returned object is acceptable.
```
network.Interface{
		Location: to.StringPtr(s.Location),
		InterfacePropertiesFormat: &network.InterfacePropertiesFormat{
			EnableAcceleratedNetworking: s.AcceleratedNetworking,
			IPConfigurations:            &ipConfigurations,
			DNSSettings:                 &dnsSettings, // <- this gets evaluated to <nil> if len(s.DNSServers) == 0
			EnableIPForwarding:          to.BoolPtr(s.EnableIPForwarding),
		},
	}, nil
```

nawazkh commented:
I see the same error as [above listed prow apidiff](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2447/pull-cluster-api-provider-azure-apidiff/1556987785515110400) on running `make apidiff` locally
```
sigs.k8s.io/cluster-api-provider-azure/azure/services/networkinterfaces
  Incompatible changes:
  - NICSpec: old is comparable, new is not
  Compatible changes:
  - NICSpec.DNSServers: added

sigs.k8s.io/cluster-api-provider-azure/azure/scope
  Incompatible changes:
  - (*MachineScope).InboundNatSpecs: changed from func(map[int32]struct{}) []sigs.k8s.io/cluster-api-provider-azure/azure.ResourceSpecGetter to func() []sigs.k8s.io/cluster-api-provider-azure/azure.ResourceSpecGetter
  .
.
.
```

Is there something else I should be taking care of since I am adding a new field(_DNSServers_) to the [NICSpec here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2447/files#diff-9e4a0ba373e2b8988e791145728267650b5b2e8a395f3f02a24e43e839081b7aR48) ?

jsturtevant commented:
I believe `pull-cluster-api-provider-azure-apidiff` isn't a required test and is expected to fail in this case.  It's a signal for the maintainers that this PR is changing the API.  I think you are all set for that test👌

nawazkh commented:
/retest-required

nawazkh commented:
/retest-required

mboersma reviewed (commented):
This is looking good. I only had one general comment which is mostly a matter of taste and can be ignored.

Could you squash the commits together?

nawazkh commented on api/v1alpha3/azuremachine_conversion.go:
```diff
@@ -58,6 +58,10 @@ func (src *AzureMachine) ConvertTo(dstRaw conversion.Hub) error {
 		dst.Spec.AdditionalCapabilities = restored.Spec.AdditionalCapabilities
 	}
 
+	if len(restored.Spec.DNSServers) != 0 {
```
Addressed the suggestion via commit bd7030074cbd4d78912eedcff78b2970254a42be
Also, you made my day. The thought process behind _"Why does Go use signed integers to index slices/arrays when there are no negative indices?"_ is beautiful. Thanks for linking it here!

nawazkh commented:
/retest

mboersma reviewed (approved):
/lgtm

CecileRobertMichon commented:
@nawazkh can you please rebase and address the merge conflict when you get a chance?

nawazkh commented:
Fixed merge conflicts and updated the PR @CecileRobertMichon @mboersma. Please take a look when you have a chance

nawazkh commented:
/retest

nawazkh commented:
> Fixed merge conflicts and updated the PR @CecileRobertMichon @mboersma. Please take a look when you have a chance

PR tests are failing because I havent updated the spec_test for the [new fields added in NICSpec](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2447/files#diff-9e4a0ba373e2b8988e791145728267650b5b2e8a395f3f02a24e43e839081b7aR51-R52)
Please hold off the review for now.

/hold

CecileRobertMichon reviewed (commented):
@nawazkh please remove the hold if this is ready

`/hold cancel`

nawazkh commented:
/hold cancel

CecileRobertMichon commented:
/lgtm
/approve

🎉

jackfrancis reviewed (approved):
lgtm
